src/Data/Unrestricted/Linear.hs

```diff
@@ -0,0 +1,175 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE RankNTypes #-}
+module Data.Unrestricted.Linear
+  ( -- * Unrestricted
+    -- $ unrestricted
+    Unrestricted(..)
+  , unUnrestricted
+    -- * Typeclasses for non-linear actions
+    -- $ comonoid
+  , Consumable(..)
+  , Dupable(..)
+  , Movable(..)
+  , void
+  , lseq
+  , dup
+  , dup2
+  , dup3
+  ) where
+-- $ unrestricted
+
+import qualified Data.Functor.Linear as Data
+import Data.Vector.Linear (V)
+import qualified Data.Vector.Linear as V
+import GHC.TypeLits
+import GHC.Types
+import qualified Unsafe.Linear as Unsafe
+
+-- | @Unrestricted a@ represents unrestricted values of type @a@ in a linear context,
+data Unrestricted a where
+  Unrestricted :: a -> Unrestricted a
+
+-- | Project an @a@ out of an @Unrestricted a@. If the @Unrestricted a@ is
+-- linear, then we get only a linear value out.
+unUnrestricted :: Unrestricted a ->. a
+unUnrestricted (Unrestricted a) = a
+
+-- $ comonoid
+
+class Consumable a where
+  consume :: a ->. ()
+
+-- | Like 'seq' but since the first argument is restricted to be of type @()@ it
+-- is consumed, hence @seqUnit@ is linear in its first argument.
+seqUnit :: () ->. b ->. b
+seqUnit () b = b
+
+-- | Like 'seq' but the first argument is restricted to be 'Consumable'. Hence the
+-- first argument is 'consume'-ed and the result consumed.
+lseq :: Consumable a => a ->. b ->. b
+lseq a b = seqUnit (consume a) b
+
+-- | The laws of @Dupable@ are dual to those of 'Monoid':
+--
+-- * @first consume (dup a) ≃ a ≃ first consume (dup a)@ (neutrality)
+-- * @first dup (dup a) ≃ (second dup (dup a))@ (associativity)
```

Comment:
We will really have to go through the documentation of this module one day: it's at best stale.